blobs: TestBlobClientWriteFile failed

blobs.TestBlobClientWriteFile [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2980306&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2980306&tab=artifacts#/) on master @ [7f82c5ef6bb3eea4147ba26e4a0a5490f2d4ff13](https://github.com/cockroachdb/cockroach/commits/7f82c5ef6bb3eea4147ba26e4a0a5490f2d4ff13):


```
=== RUN   TestBlobClientWriteFile
--- FAIL: TestBlobClientWriteFile (0.02s)
=== RUN   TestBlobClientWriteFile/write-outside-extern-dir
    client_test.go:270: EOF
    --- FAIL: TestBlobClientWriteFile/write-outside-extern-dir (0.00s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestBlobClientWriteFile PKG=./pkg/blobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</p>
</details>
/cc @cockroachdb/bulk-io @g3orgia
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBlobClientWriteFile.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

blobs.TestBlobClientWriteFile [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2987376&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2987376&tab=artifacts#/) on master @ [5595273db13444939198b2ff2c5e8a208935e4d8](https://github.com/cockroachdb/cockroach/commits/5595273db13444939198b2ff2c5e8a208935e4d8):


```
=== RUN   TestBlobClientWriteFile
--- FAIL: TestBlobClientWriteFile (0.03s)
=== RUN   TestBlobClientWriteFile/write-outside-extern-dir
    client_test.go:270: EOF
    --- FAIL: TestBlobClientWriteFile/write-outside-extern-dir (0.00s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestBlobClientWriteFile PKG=./pkg/blobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/bulk-io @g3orgia
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBlobClientWriteFile.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

blobs.TestBlobClientWriteFile [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2987459&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2987459&tab=artifacts#/) on master @ [5595273db13444939198b2ff2c5e8a208935e4d8](https://github.com/cockroachdb/cockroach/commits/5595273db13444939198b2ff2c5e8a208935e4d8):


```
=== RUN   TestBlobClientWriteFile
--- FAIL: TestBlobClientWriteFile (0.04s)
=== RUN   TestBlobClientWriteFile/write-outside-extern-dir
    client_test.go:270: EOF
    --- FAIL: TestBlobClientWriteFile/write-outside-extern-dir (0.00s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestBlobClientWriteFile PKG=./pkg/blobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/bulk-io @g3orgia
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBlobClientWriteFile.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/blobs/client.go
+++ b/pkg/blobs/client.go
@@ -114,7 +114,8 @@ func (w *streamWriter) CloseWithError(err error) error {
 		return w.Close()
 	}
 	w.cancel()
-	return nil
+	_, err = w.s.CloseAndRecv()
+	return err
 }
 
 func (c *remoteClient) Writer(ctx context.Context, file string) (WriteCloserWithError, error) {

--- a/pkg/blobs/client_test.go
+++ b/pkg/blobs/client_test.go
@@ -258,7 +258,7 @@ func TestBlobClientWriteFile(t *testing.T) {
 					return err
 				}
 				if _, err := io.Copy(w, bytes.NewReader(byteContent)); err != nil {
-					return errors.CombineErrors(err, w.CloseWithError(err))
+					return errors.CombineErrors(w.CloseWithError(err), err)
 				}
 				return w.Close()
 			}()

--- a/pkg/sql/copy_file_upload.go
+++ b/pkg/sql/copy_file_upload.go
@@ -169,7 +169,7 @@ func CopyInFileStmt(destination, schema, table string) string {
 func (f *fileUploadMachine) run(ctx context.Context) error {
 	err := f.c.run(ctx)
 	if err != nil {
-		err = errors.CombineErrors(err, f.w.CloseWithError(err))
+		err = errors.CombineErrors(f.w.CloseWithError(err), err)
 	} else {
 		err = f.w.Close()
 	}

--- a/pkg/storage/cloud/cloud_io.go
+++ b/pkg/storage/cloud/cloud_io.go
@@ -305,7 +305,7 @@ func WriteFile(ctx context.Context, dest ExternalStorage, basename string, src i
 	}
 	if _, err := io.Copy(w, src); err != nil {
 		closeErr := w.CloseWithError(err)
-		return errors.CombineErrors(err, closeErr)
+		return errors.CombineErrors(closeErr, err)
 	}
 	return errors.Wrap(w.Close(), "closing object")
 }

--- a/pkg/storage/cloud/external_storage.go
+++ b/pkg/storage/cloud/external_storage.go
@@ -73,9 +73,19 @@ type ExternalStorage interface {
 	// This can be leveraged for an existence check.
 	ReadFileAt(ctx context.Context, basename string, offset int64) (io.ReadCloser, int64, error)
 
-	// Writer returns a writer for the requested name. The returned writer must
-	// be closed by the caller to complete the write, or can be closed with an
-	// error which may, depending on the implementation, abort the write.
+	// Writer returns a writer for the requested name.
+	//
+	// A Writer *must* be closed via either Close or CloseWithError, and if
+	// closing returns a non-nil error, that error should be handled or reported
+	// to the user (likely using errors.CombineErrors(closeErr, reasonErr). This
+	// is because an implementation may wait or buffer written data until Close
+	// or a Write call may return io.EOF if the writer's stream is closed due to
+	// an error that will be reported by Close.
+	//
+	// The CloseWithError(err) alternative to Close(), if passed a non-nil error
+	// may elect to abort the operation, skip flushing any buffers or otherwise
+	// end the operation with the minimum additional work in the event the
+	// caller no longer wishes to complete it, e.g. due to some other error.
 	Writer(ctx context.Context, basename string) (WriteCloserWithError, error)
 
 	// ListFiles returns files that match a globs-style pattern. The returned

